src/vm/checker/typecheck/interface.rs

```diff
@@ -0,0 +1,226 @@
+            args.push(ContractInterfaceFunctionArg { 
+                name: fnArg.name.to_string(), 
+                type_f: ContractInterfaceAtomType::from_type_signature(&fnArg.signature)
+            });
+        }
+        args
+    }
+}
+
+#[derive(Debug, Serialize)]
+pub struct ContractInterfaceFunctionOutput {
+    #[serde(rename = "type")]
+    pub type_f: ContractInterfaceAtomType,
+}
+
+#[derive(Debug, Serialize)]
+pub struct ContractInterfaceFunction {
+    pub name: String,
+    pub access: ContractInterfaceFunctionAccess,
+    pub args: Vec<ContractInterfaceFunctionArg>,
+    pub outputs: ContractInterfaceFunctionOutput,
+}
+
+impl ContractInterfaceFunction {
+    pub fn from_map(map: &BTreeMap<String, FunctionType>, access: ContractInterfaceFunctionAccess) -> Vec<ContractInterfaceFunction> {
+        map.iter().map(|(name, function_type)| {
+            ContractInterfaceFunction {
+                name: name.to_string(),
+                access: access.to_owned(),
+                outputs: ContractInterfaceFunctionOutput { 
+                    type_f: match function_type {
+                        FunctionType::Fixed(_, fnType) => {
+                            ContractInterfaceAtomType::from_type_signature(&fnType)
+                        },
+                        FunctionType::Variadic(_, _) => panic!("Contract functions should never have a variadic return type!"),
+                        FunctionType::UnionArgs(_, _) => panic!("Contract functions should never have a union return type!"),
+                    }
+                },
+                args: match function_type {
+                    FunctionType::Fixed(fnArgs, _) => {
+                        ContractInterfaceFunctionArg::from_function_args(&fnArgs)
+                    },
+                    FunctionType::Variadic(_, _) => panic!("Contract functions should never have variadic arguments!"),
+                    FunctionType::UnionArgs(_, _) => panic!("Contract functions should never have union arguments!"),
+                }
+            }
+        }).collect()
+    }
+}
+
+#[derive(Debug, Serialize, Clone)]
+pub enum ContractInterfaceVariableAccess {
+    constant,
+    variable,
+}
+
+#[derive(Debug, Serialize)]
+pub struct ContractInterfaceVariable { 
+    pub name: String,
+    #[serde(rename = "type")]
+    pub type_f: ContractInterfaceAtomType,
+    pub access: ContractInterfaceVariableAccess,
+}
+
+impl ContractInterfaceVariable {
+    pub fn from_map(map: &BTreeMap<String, TypeSignature>, access: ContractInterfaceVariableAccess) -> Vec<ContractInterfaceVariable> {
+        map.iter().map(|(name, type_sig)| {
+            ContractInterfaceVariable {
+                name: name.to_string(),
+                access: access.to_owned(),
+                type_f: ContractInterfaceAtomType::from_type_signature(type_sig),
+            }
+        }).collect()
+    }
+}
+
+#[derive(Debug, Serialize)]
+pub struct ContractInterfaceMap {
```

Comment:
I don't think this struct is quite right -- a map is something like a

Key of type tuple
and
Value of type tuple

Neither the key or the value generally has a name.

e.g.

`(define-map foo-bar ((k1 int) (k2 bool)) ((v1 int) (v2 bool)))`

is a map from keys of type `(tuple (k1 int) (k2 bool))` to values of type `(tuple (v1 int) (v2 bool))`

I think the `ContractInterfaceMap::from_map` function is only grabbing the first entry in the tuple.

Reply:
@kantai Updated the representation. Does it look correct now?

Reply:
Yep, it looks good to me.